File: sketch/give_advice.cpp
#include <iostream>

/** 场景：邹忌讽齐王纳谏
 * 邹忌修八尺有余，而形貌昳丽。
 * 朝服衣冠，窥镜，谓其妻曰：“我孰与城北徐公美？” 其妻曰：“君美甚，徐公何能及君也？”
 * 城北徐公，齐国之美丽者也。
 * 忌不自信，而复问其妾曰：“吾孰与徐公美？”妾曰：“徐公何能及君也？”
 * 旦曰，客从外来，与坐谈，问之：“吾与徐公孰美？”客曰：“徐公不若君之美也。”
 * 明日，徐公来，孰视之，自以为不如；
 * 窥镜而自视，又弗如远甚。
 * 
 * 暮寝而思之，曰：“吾妻之美我者，私我也；妾之美我者，畏我也；客之美我者，欲有求于我也。”
 */

// 定义类和方法
struct Person {
    Person(const char* name, unsigned char good_looking = 6) {
        this->name = name;
        this->颜值 = good_looking;
    }

    // 人会臭美，问别人谁更美
    void 问(const Person* 回答者, const Person* 比较者) const {
        printf("%s: 吾孰与%s美?\n", this->name, 比较者->name);
        回答者->曰(this, 比较者);
    }
    
    // 人会奉承，见人说人话 见鬼说鬼话
    void 曰(const Person* 提问者, const Person* 比较者) const {
        printf("%s: 君美甚，%s何能及君也!\n", this->name, 比较者->name);
    }

    // 人会自省，但多少有点偏心，颜值相同时也会觉得自己更美
    void 自省(const Person* 比较者) const {
        if (this->颜值 < 比较者->颜值) {
            printf("[%s窥镜而自视: 吾诚知不如%s美!]\n", this->name, 比较者->name);
        } else {
            printf("[%s窥镜而自视: 吾诚知%s不若吾之美也!]\n", this->name, 比较者->name);
        }
    }

    const char* name;
    unsigned char 颜值;
};

int main(int 客人数量, char* 客人列表[]) {
    printf("邹忌讽齐王纳谏: %s\n", 客人列表[0]);

    // 角色就绪
    Person 邹忌("邹忌", 8), 徐公("徐公", 10);
    Person 妻("妻"), 妾("妾");

    // 还原情景
    邹忌.问(&妻, &徐公);
    邹忌.问(&妾, &徐公);

    for (int i = 1; i < 客人数量; i++) {
        auto 客 = new Person(客人列表[i]);

        printf("旦日, %s从外来, 与座谈.\n", 客->name);
        邹忌.问(客, &徐公);
        邹忌.自省(客);

        delete 客;
    }

    printf("明日, %s来, 孰视之.\n", 徐公.name);
    邹忌.自省(&徐公);

    return 0;
}

File: sketch/give_advice.py
import random

# 场景：邹忌讽齐王纳谏

# 定义类和方法
class Person(object):
    def __init__(self, name, good_looking = random.randint(1, 10)):
        self.name = name
        self.颜值 = good_looking
        print(f"{name}已就绪")

    # 人会臭美，喜欢问别人谁颜值高
    def 问(self, 回答者, 比较者):
        print(f"{self.name}: 吾孰与{比较者.name}美？")
        回答者.曰(self, 比较者)
    
    # 人会奉承，见人说人话 见鬼说鬼话
    def 曰(self, 提问者, 比较者):
        print(f"{self.name}: 君美甚，{比较者.name}何能及君也！")

    # 人会自省，但多少有点偏心，即使颜值相同也会觉得自己更美
    def 自省(self, 比较者):
        if self.颜值 < 比较者.颜值:
            print(f"[{self.name}窥镜而自视：吾诚知不如{比较者.name}美！]")
        else:
            print(f"[{self.name}窥镜而自视：吾诚知{比较者.name}何能及吾也！]")

# 角色就绪
邹忌 = Person("邹忌", 8)
徐公 = Person("徐公", 10)

妻 = Person("妻")
妾 = Person("妾")
客 = Person("客")

# 完整还原情景
邹忌.问(妻, 徐公)
邹忌.问(妾, 徐公)
邹忌.问(客, 徐公)
邹忌.自省(徐公)
